### PR TITLE
feat: Move safe IP address handling code from edx-platform

### PR DESCRIPTION
**Pull request by timmc-edx**
This will allow using the code in all IDAs.

Part of <https://github.com/openedx/edx-django-utils/issues/241>.
Code will be removed from edx-platform in <https://github.com/openedx/edx-platform/pull/31000>.

----

Commits best reviewed separately.

----

**Merge checklist:**
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are squashed

### CODE REVIEW

File: edx_django_utils/ip/README.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Request Utils
```
Comment (robrap) [inline edx_django_utils/ip/README.rst:1]: IP Utils?

File: edx_django_utils/ip/internal/ip.py [hunk 2]
```diff
@@ -357,15 +357,16 @@ def _get_client_ips_via_trusted_header(request):
         if closest_client_ip := _get_trusted_header_ip(request, header_name, index):
             # The equality check in this predicate is why we use parsed IP
             # addresses -- ::1 should compare as equal to 0:0:0:0:0:0:0:1.
-            external_chain = _remove_tail(full_chain, lambda ip: ip != closest_client_ip)  # pylint: disable=cell-var-from-loop
+            # pylint: disable=cell-var-from-loop
+            external_chain = _remove_tail(full_chain, lambda ip: ip != closest_client_ip)
```
Comment (robrap) [inline edx_django_utils/ip/internal/ip.py:201]: Nit: Why this change? Did it also have line too long? You could disable that as well. You can leave this, but I like getting this stuff off to the side.

File: edx_django_utils/ip/internal/tests/test_ip.py [hunk 3]
```diff
@@ -132,7 +132,8 @@ def test_get_client_ips_via_xff(self, xff, remote_addr, expected_strs):
         request_meta = {k: v for k, v in request_meta.items() if v is not None}
         self.request.META = request_meta
 
-        assert [str(ip) for ip in ip._get_client_ips_via_xff(self.request)] == expected_strs  # pylint: disable=protected-access
+        # pylint: disable=protected-access
+        assert [str(ip) for ip in ip._get_client_ips_via_xff(self.request)] == expected_strs
```
Comment (robrap) [inline edx_django_utils/ip/internal/tests/test_ip.py:136]: Nit: Same comment about pylint positioning.

**Review by robrap — COMMENTED** [review]
1. Can you update the PR description so it stands on its own? Maybe just repeat the title, if you don't feel the need to say more. (This is personal preference.)
2. What is the plan for test coverage? Should we do it as part of this PR, since it will be library code that has a larger scope? If not, we could add pragmas with a comment pointing to the ticket to complete this, like a temporary amnesty.
3. I did not pay much attention to code and tests that seemed to just be copied over from the original.

File: edx_django_utils/ip/internal/ip.py (hunk 2 above)
Comment (timmc-edx) [reply edx_django_utils/ip/internal/ip.py:201]: Yeah, pycodestyle E501 line-too-long, and I was having trouble figuring out how to disable _just_ that warning (rather than using `noqa`). I can take another look to see if there's a way to do that,

Comment (timmc-edx) [reply edx_django_utils/ip/internal/ip.py:201]: OK, apparently `noqa` has to appear _first_ in the comment, and doesn't accept things like `E501` -- it just entirely disables all errors of any sort on that line.

File: edx_django_utils/ip/README.rst (hunk 1 above)
Comment (timmc-edx) [reply edx_django_utils/ip/README.rst:1]: Ah yes, thank you -- that was an artifact from an earlier version of this branch when I was considering having this be a broader "request utils" module.

File: edx_django_utils/ip/internal/ip.py (hunk 2 above)
Comment (robrap) [reply edx_django_utils/ip/internal/ip.py:201]: Are you saying the recommendation to use `# noqa: E501, pylint: disable=line-too-long` doesn't work?

Comment (robrap) [reply edx_django_utils/ip/internal/ip.py:201]: I’m just curious, but feel free to leave this on two lines. I don’t mean to get us lost in this. Also why I wish there was a doc for this, if it could be solved once.

**Comment by timmc-edx** [comment]
I don't like repeating the PR title in the body. I can link in the pending edx-platform PR, though. Do you think there's more that needs to be said beyond that?

**Comment by timmc-edx** [comment]
I'm going to try to improve coverage and add pragmas where it doesn't seem worth it.

**Comment by robrap** [comment]
> I don't like repeating the PR title in the body. I can link in the pending edx-platform PR, though. Do you think there's more that needs to be said beyond that?

Nope. We just have different preferences. Your PR. Go with yours. 🙂

File: edx_django_utils/ip/internal/ip.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    Return the list of IPs the request is carrying on this header, which is
+    expected to be comma-delimited if it contains more than one. Response
+    may be an empty list for missing or empty header. List items may not be
+    valid IPs.
+    """
+    if not header_name:
+        return []
+
+    field_name = 'HTTP_' + header_name.replace('-', '_').upper()
+    header_value = request.META.get(field_name, '').strip()
```
Comment (timmc-edx) [inline edx_django_utils/ip/internal/ip.py:28]: I'm not sure why I'm not using `request.headers` here, and instead seem to have duplicated the logic that that uses. I wonder if it's possible that this was to work around the "request" objects we get sometimes being `django.http.request.HttpRequest` and sometimes being a DRF wrapper? But the DRF wrapper does proxy through any "missing" attributes via a dynamic `__getattr__`, so `headers` is available. Maybe I thought it wasn't.

Anyway, I'm going to see if I can just replace this code with `request.headers.get(field_name, '').strip()`.

**Comment by timmc-edx** [comment]
Rewrote branch into two commits: One that's basically a straight copy, and one that makes new changes.

**Review by robrap — APPROVED** [review]

**Comment by timmc-edx** [comment]
(rebased onto master)

**Comment by timmc-edx** [comment]
(adjusted changelog date)
